File: nucleo/gih_nucleo/viabilidade.py
"""Decide se a campanha tem solução antes de procurar uma (H52, RF31, RN07).

Com as cotas em contagem (RN11), a pergunta tem resposta exata, e não
heurística. Toda ação custa pelo menos o que custa a mais barata, e todo
parceiro pode recebê-la; então os mínimos cabem na campanha se, e somente se,
o **menor conjunto de parceiros que os cumpre** couber no máximo de ações e no
orçamento pagando a ação mais barata para cada um.

O menor conjunto se monta em dois passos:

1. Em cada categoria, o mínimo é preenchido primeiro com parceiros da cauda
   longa: cada um conta para as duas cotas ao mesmo tempo.
2. O que ainda faltar de cauda longa sai das categorias que têm folga no
   máximo, ou de quem não tem categoria confirmada.

Nenhum conjunto que cumpra as cotas é menor que esse. E ele é também o começo
da solução inicial do genético, que por isso nasce viável (ADR-011).

Quando falha, a recusa diz qual restrição e quanto falta: `exigido` contra
`disponivel`. O texto para a pessoa é montado pela API, que sabe o nome da
categoria e formata dinheiro; o núcleo só conhece índices e centavos.
"""
from __future__ import annotations

from dataclasses import dataclass

from gih_nucleo.problema import Instancia

# Qual restrição impede a campanha, na ordem em que são conferidas.
COTA_CATEGORIA = "cota_categoria"  # o mínimo passa do máximo
ELEGIVEIS_CATEGORIA = "elegiveis_categoria"  # a categoria não tem parceiros para o mínimo
CAUDA_LONGA = "cauda_longa"  # não há cauda longa que caiba nas cotas
MAXIMO_ACOES = "maximo_acoes"  # os mínimos somam mais que K
ORCAMENTO = "orcamento"  # os mínimos não cabem no orçamento nem com a ação mais barata


@dataclass(frozen=True)
class Inviabilidade:
    restricao: str
    exigido: int  # contagem, ou centavos quando a restrição é o orçamento
    disponivel: int
    categoria: int | None = None

    @property
    def falta(self) -> int:
        return self.exigido - self.disponivel


class Inviavel(Exception):
    """A campanha não tem solução; `motivo` diz por quê. Nenhum plano é devolvido (RN07)."""

    def __init__(self, motivo: Inviabilidade) -> None:
        super().__init__(motivo.restricao)
        self.motivo = motivo


def _diagnosticar(inst: Instancia) -> tuple[list[int], Inviabilidade | None]:
    barata = inst.acao_mais_barata

    # Dentro de cada grupo, os de maior ganho com a ação mais barata primeiro:
    # o conjunto mínimo já sai com os melhores candidatos, e o desempate pelo
    # índice o deixa igual em qualquer versão.
    def prioridade(i: int) -> tuple[int, int]:
        return (-inst.ganho[i][barata], i)

    membros: list[list[int]] = [[] for _ in range(inst.categorias)]
    cauda_sem_categoria = []
    for i, k in enumerate(inst.categoria):
        if k >= 0:
            membros[k].append(i)
        elif inst.cauda[i]:
            cauda_sem_categoria.append(i)

    minimos, maximos = inst.minimo_categoria, inst.maximo_categoria
    for k in range(inst.categorias):
        if minimos[k] > maximos[k]:
            return [], Inviabilidade(COTA_CATEGORIA, minimos[k], maximos[k], k)
    for k in range(inst.categorias):
        if minimos[k] > len(membros[k]):
            return [], Inviabilidade(ELEGIVEIS_CATEGORIA, minimos[k], len(membros[k]), k)

    escolhidos: list[int] = []
    cobertura = 0  # quanto da cota da cauda os mínimos de categoria já cobrem
    reserva_cauda = list(cauda_sem_categoria)
    for k in range(inst.categorias):
        ordem = sorted(membros[k], key=prioridade)
        da_cauda = [i for i in ordem if inst.cauda[i]]
        fora_da_cauda = [i for i in ordem if not inst.cauda[i]]
        usados = min(minimos[k], len(da_cauda))
        escolhidos += da_cauda[:usados]
        escolhidos += fora_da_cauda[: minimos[k] - usados]
        cobertura += usados
        # A cauda que sobra nesta categoria só completa a cota até o máximo dela.
        reserva_cauda += da_cauda[usados:][: maximos[k] - minimos[k]]

    falta_cauda = max(0, inst.minimo_cauda - cobertura)
    if falta_cauda > len(reserva_cauda):
        return [], Inviabilidade(CAUDA_LONGA, inst.minimo_cauda, cobertura + len(reserva_cauda))
    escolhidos += sorted(reserva_cauda, key=prioridade)[:falta_cauda]

    if len(escolhidos) > inst.maximo_acoes:
        return [], Inviabilidade(MAXIMO_ACOES, len(escolhidos), inst.maximo_acoes)
    custo = len(escolhidos) * inst.custo[barata]
    if custo > inst.orcamento:
        return [], Inviabilidade(ORCAMENTO, custo, inst.orcamento)
    return sorted(escolhidos), None


def verificar_viabilidade(inst: Instancia) -> Inviabilidade | None:
    """`None` quando a campanha tem solução; senão, a primeira restrição que a impede."""
    return _diagnosticar(inst)[1]


def conjunto_minimo(inst: Instancia) -> list[int]:
    """Os parceiros do menor conjunto que cumpre as cotas; levanta `Inviavel` se não houver."""
    escolhidos, motivo = _diagnosticar(inst)
    if motivo is not None:
        raise Inviavel(motivo)
    return escolhidos

File: nucleo/cpp/viabilidade.cpp
// A decisão exata de viabilidade e o menor conjunto que cumpre as cotas —
// reproduz `gih_nucleo/viabilidade.py` (`_diagnosticar`). O porquê de a conta ser
// exata está lá e na `docs/07` §4.1.
#include <algorithm>

#include "nucleo.hpp"

namespace gih {

bool diagnosticar(const Instancia& inst, std::vector<int>& escolhidos, Inviabilidade& motivo) {
    escolhidos.clear();
    const int barata = inst.acao_mais_barata();

    // Os de maior ganho com a ação mais barata primeiro; no empate, o menor índice.
    auto prioridade = [&](int i, int j) {
        const std::int64_t gi = inst.ganho_de(i, barata), gj = inst.ganho_de(j, barata);
        if (gi != gj) return gi > gj;
        return i < j;
    };

    std::vector<std::vector<int>> membros(static_cast<std::size_t>(inst.categorias));
    std::vector<int> reserva_cauda;  // começa com a cauda sem categoria
    for (int i = 0; i < inst.parceiros; ++i) {
        const int k = inst.categoria[i];
        if (k >= 0) {
            membros[k].push_back(i);
        } else if (inst.cauda[i]) {
            reserva_cauda.push_back(i);
        }
    }

    for (int k = 0; k < inst.categorias; ++k) {
        if (inst.minimo_categoria[k] > inst.maximo_categoria[k]) {
            motivo = {"cota_categoria", inst.minimo_categoria[k], inst.maximo_categoria[k], k};
            return false;
        }
    }
    for (int k = 0; k < inst.categorias; ++k) {
        const auto tamanho = static_cast<std::int64_t>(membros[k].size());
        if (inst.minimo_categoria[k] > tamanho) {
            motivo = {"elegiveis_categoria", inst.minimo_categoria[k], tamanho, k};
            return false;
        }
    }

    std::int64_t cobertura = 0;
    for (int k = 0; k < inst.categorias; ++k) {
        std::vector<int> ordem = membros[k];
        std::sort(ordem.begin(), ordem.end(), prioridade);
        std::vector<int> da_cauda, fora_da_cauda;
        for (int i : ordem) (inst.cauda[i] ? da_cauda : fora_da_cauda).push_back(i);

        const std::int64_t minimo = inst.minimo_categoria[k];
        const std::int64_t usados = std::min<std::int64_t>(minimo, static_cast<std::int64_t>(da_cauda.size()));
        escolhidos.insert(escolhidos.end(), da_cauda.begin(), da_cauda.begin() + usados);
        escolhidos.insert(escolhidos.end(), fora_da_cauda.begin(), fora_da_cauda.begin() + (minimo - usados));
        cobertura += usados;

        // A cauda que sobra nesta categoria só completa a cota até o máximo dela.
        const std::int64_t sobra = static_cast<std::int64_t>(da_cauda.size()) - usados;
        const std::int64_t cabem = std::min(sobra, inst.maximo_categoria[k] - minimo);
        reserva_cauda.insert(reserva_cauda.end(), da_cauda.begin() + usados, da_cauda.begin() + usados + cabem);
    }

    const std::int64_t falta_cauda = std::max<std::int64_t>(0, inst.minimo_cauda - cobertura);
    const auto reserva = static_cast<std::int64_t>(reserva_cauda.size());
    if (falta_cauda > reserva) {
        motivo = {"cauda_longa", inst.minimo_cauda, cobertura + reserva, SEM_CATEGORIA};
        return false;
    }
    std::sort(reserva_cauda.begin(), reserva_cauda.end(), prioridade);
    escolhidos.insert(escolhidos.end(), reserva_cauda.begin(), reserva_cauda.begin() + falta_cauda);

    const auto total = static_cast<std::int64_t>(escolhidos.size());
    if (total > inst.maximo_acoes) {
        motivo = {"maximo_acoes", total, inst.maximo_acoes, SEM_CATEGORIA};
        return false;
    }
    const std::int64_t custo = total * inst.custo[barata];
    if (custo > inst.orcamento) {
        motivo = {"orcamento", custo, inst.orcamento, SEM_CATEGORIA};
        return false;
    }
    std::sort(escolhidos.begin(), escolhidos.end());
    return true;
}

}  // namespace gih
